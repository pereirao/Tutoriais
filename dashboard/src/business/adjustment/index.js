// @flow

export type Adjustment = {
  id: number,
  created_at: string, // rails timestamp
  amount: number,
  credit: boolean,
  reason: string,
  description: string
};

export * as adjustment_actions from './actions';
export * as adjustment_epics from './epics';
export * as adjustment_helpers from './helpers';
export { default as adjustmentReducer } from './reducer';
export { default as adjustment_selectors } from './selectors';
