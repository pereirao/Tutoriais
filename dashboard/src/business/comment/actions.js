// @flow

export const fetchComments = (order_id: string, options: {from_order_detail: boolean} = {from_order_detail: true}) => ({
  type: 'COMMENT:FETCH',
  payload: { order_id },
  meta: { order_id, ...options }
});

export const addComment = (order_id: string, body) => ({
  type: 'COMMENT:ADD',
  payload: { order_id, body },
  meta: { order_id }
});
