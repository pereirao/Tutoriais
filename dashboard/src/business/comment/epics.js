// @flow

import {filter, switchMap} from 'rxjs/operators';
import {api} from '../networking';
import createActionsForRequest from '../utils/create_actions_for_request';

export const fetchComments = (action$: Object) => {
    const comment_response_action$ = action$
        .pipe(
            filter(action => action.type === 'COMMENT:FETCH'),
            switchMap(action => {
                return createActionsForRequest(api.fetchComments(action.payload.order_id), action.type, action.meta);
            })
        );

    return comment_response_action$;
};

export const addComment = (action$: Object) => {
    const comment_response_action$ = action$
        .pipe(
            filter(action => action.type === 'COMMENT:ADD'),
            switchMap(action => {
                return createActionsForRequest(api.addComment(action.payload.order_id, action.payload.body), action.type, action.meta);
            })
        );
    

    return comment_response_action$;
};
