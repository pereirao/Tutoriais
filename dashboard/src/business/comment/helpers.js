// @flow

import _ from 'lodash';
import moment from 'moment';
import I18n from '../../localization';
import { date_helpers } from '../../views/utils/helpers';
import type { Exception } from '../order';
import type { Comment } from './index';

const { formatTimestamp, minutesAgo } = date_helpers;

const MINIBAR_EMAIL_DOMAIN_SUFFIX = 'minibardelivery.com';
export const SENT_TO_CUSTOMER_PREFIX = 'SENT TO CUSTOMER:';

// TODO: create display reason through i18n
export const commentNote = (comment: Comment) => comment.note;
export const commentAttribution = (comment: Comment) => (
  `${formattedAuthor(comment)} - ${formatTimestamp(comment.created_at)}`
);
export const isFromMinibar = (comment: Comment) => {
  if (comment.posted_as){
    return comment.posted_as === 'minibar';
  }
  return comment.author.email && comment.author.email.endsWith(MINIBAR_EMAIL_DOMAIN_SUFFIX);
};

export const isSupplier = (comment: Comment) => comment && comment.posted_as === 'supplier';

// e.g. Brian Cooper (Minibar CX) or Pratik Shah (employee)
export const formattedAuthor = (comment: Comment) => {
  const options = {name: comment.author.name};
  return isFromMinibar(comment) ? I18n.t('ui.body.comment_author.cx', options) : I18n.t('ui.body.comment_author.employee', options);
};

export const commentTimelineEntries = (comments: Array<Comment>) => (
  comments.map(comment => ({time: comment.created_at, type: 'comment', meta: {name: formattedAuthor(comment)} }))
);

export const sentToCustomer = (comment: Comment) => comment.note && comment.note.startsWith(SENT_TO_CUSTOMER_PREFIX);
export const trimNotePrefix = (note: string) => note.split(': ')[1];

export const commentNotification = (comment: Comment, order_id: string) => {
  const notification = {id: `comment_${order_id}`, order_id, notification_type: 'comment' };
  return notification;
};

export const createdAt = (comment: Comment) => comment && formatTimestamp(moment.parseZone(comment.created_at));
export const minutesSinceCreated = (comment: Comment) => {
  return minutesAgo(moment.parseZone(comment.created_at)).toString();
};

export const exceptionComment = (exception: Exception) => {
  const exception_items = _.map(exception.metadata, (value, key) => {
    const key_text = I18n.t(`ui.body.report_problem.${exception.type}.${key}`);
    return `${key_text}: ${value}`;
  });
  return exception_items.join('\n');
};
