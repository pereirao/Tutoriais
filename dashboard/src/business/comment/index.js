// @flow
export type Author = {
  email: string
};

export type Comment = {
  id: number,
  created_at: string, // rails timestamp
  note: string,
  author: Author,
  posted_as: string
};

export * as comment_actions from './actions';
export * as comment_epics from './epics';
export * as comment_helpers from './helpers';
export { default as commentReducer } from './reducer';
export { default as comment_selectors } from './selectors';
