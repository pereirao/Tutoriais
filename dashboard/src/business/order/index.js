// @flow

export type OrderTags = {
  gift: boolean,
  out_of_hours: boolean,
  vip: boolean,
  corporate: boolean,
  scheduled: boolean,
  new_customer: boolean
};

export type BillingInfo = {
  name: string,
  card_type: string,
  card_last4: string
};

export type Address = {
  address1: string,
  address2: string,
  city: string,
  state: string,
  zip_code: string,
  coords: {lat: number, lng: number}
};

export type PickupDetail = {
  id: string,
  name: string,
  phone: string
};

export type Amounts = {
  total: number,
  subtotal: number,
  tax: number,
  tip: number,
  minibar_promos: number,
  store_discounts: number,
  discounts?: number, // FIXME: optional until sunset
  delivery_fee: number,
};

export type DeliveryService = {
  id: number,
  name: string,
  email: string,
  created_at: string,
  updated_at: string,
  minimal_delivery_time: number
}

export type DeliveryMethodType = 'on_demand' | 'shipped' | 'pickup';

export type DeliveryMethod = {
  type: DeliveryMethodType,
  expectation: string
};

export type AttributeFilters = 'vip' | 'gift' | 'scheduled' | 'exception';

export type OrderFilters = {
  date_range: {
    start: string, // iso-8601 timestamp
    end: string // iso-8601 timestamp
  },
  delivery_method_types: Array<DeliveryMethodType>,
  attributes: Array<AttributeFilters>
}

export type ExceptionMetadata = {
  order_item_sku?: string,
  replacement_item?: string,
  replacement_price?: string,
  correct_price?: string,
  incorrect_listing?: string,
  incomplete_address?: string,
  suspected_fraud?: string,
  failed_delivery?: string,
  driver_unavailable_why?: string,
  other?: string,
};

export type ExceptionType = 'out_of_stock' |
  'incorrect_pricing' |
  'incorrect_listing' |
  'incomplete_address' |
  'suspected_fraud' |
  'driver_unavailable' |
  'failed_delivery' |
  'other';

export type Exception = {
  type: ExceptionType,
  description: string,
  metadata: ExceptionMetadata
};

export const PROBLEM_TYPES = [
  'out_of_stock',
  'incorrect_pricing',
  'incorrect_listing',
  'incomplete_address',
  'suspected_fraud',
  'driver_unavailable',
  'failed_delivery',
  'other'
];

export type OrderListName = 'all' | 'active' | 'completed' | 'future';

export type RecipientInfo = { short_name: string, long_name: string, phone: string };

export type TimelineEntryType = 'created_at'
  | 'confirmed_at'
  | 'en_route_at'
  | 'delivered_at'
  | 'canceled_at'
  | 'comment';

export type TimelineEntry = {
  time: string,
  type: TimelineEntryType,
  meta?: { name?: string, driverName?: string }
}

export type OrderState = 'paid'
  | 'confirmed'
  | 'scheduled'
  | 'canceled'
  | 'exception'
  | 'en_route'
  | 'delivered';

export type OrderItem = {
  id: number,
  sku: string,
  name: string,
  item_size: string,
  unit_price: number,
  quantity: number
};

export type Order = {
  id: number,
  number: number,
  created_at: string, // rails timestamp
  confirmed_at: string, // rails timestamp
  canceled_at: string, // rails timestamp
  delivered_at: string, // rails timestamp
  scheduled_for: string, // rails timestamp
  scheduled_for_end: string, // rails timestamp
  order_time: string, // rails timestamp
  birthdate: string,
  notes: string,
  delivery_method: DeliveryMethod,
  state: OrderState,
  gift_message: string,
  recipient_info: RecipientInfo,
  customer_name: string,
  billing: BillingInfo,
  address?: Address,
  pickup_detail?: PickupDetail,
  amounts: Amounts,
  type_tags: OrderTags,
  receipt_url: string,
  order_items: Array<OrderItem>,
  comment_ids: Array<number>, // NOTE: shouldn't be hydrated
  order_adjustment_ids: Array<number>, // NOTE: shouldn't be hydrated
  driver_id: number,
  driver?: any,
  using_delivery_service: boolean,
  delivery_service: DeliveryService
};

export * as order_actions from './actions';
export * as order_helpers from './helpers';
export * as order_epics from './epics';
export { default as orderReducer } from './reducer';
export { default as order_selectors } from './selectors';
