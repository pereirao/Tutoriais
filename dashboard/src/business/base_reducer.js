//@flow

import {combineReducers} from 'redux';
import {adjustmentReducer} from './adjustment';
import {commentReducer} from './comment';
import {employeeReducer} from './employee';
import {notificationReducer} from './notification';
import {notificationMethodReducer} from './notification_method';
import {orderReducer} from './order';
import {reportReducer} from './report';
import {uiReducer} from './ui';
import {sessionReducer} from './session';
import {variantReducer} from './variant';
import {settingsReducer} from './settings';
import substituteReducer from './substitute/substitute.dux';
import productReducer from './product/product.dux';
import {connectRouter} from "connected-react-router";

const appReducer = (history) => combineReducers({
  adjustment: adjustmentReducer,
  comment: commentReducer,
  employee: employeeReducer,
  notification: notificationReducer,
  notification_method: notificationMethodReducer,
  order: orderReducer,
  report: reportReducer,
  ui: uiReducer,
  router: connectRouter(history),
  session: sessionReducer,
  variant: variantReducer,
  settings: settingsReducer,
  substitute: substituteReducer,
  product: productReducer
});

const rootReducer = (history) => (state, action) => {
  let app_state = state;
  if (action.type === 'SESSION:RESET_STATE') {
    app_state = undefined;
  }
  return appReducer(history)(app_state, action);
};

export default rootReducer;
