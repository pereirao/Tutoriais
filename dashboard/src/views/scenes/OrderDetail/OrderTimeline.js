// @flow

import React, {PureComponent} from 'react';
import _ from 'lodash';
import moment from 'moment';
import {connect} from 'react-redux';
import * as Ent from '@minibar/store-business/src/utils/ent';
import classNames from 'classnames/bind';
import style_sheet from './OrderDetail.module.css';
import {date_helpers} from '../../utils/helpers';
import MBIcon from '../../elements/MBIcon';
import type {Order, TimelineEntry} from '../../../business/order';
import {order_helpers} from '../../../business/order';
import {comment_helpers} from '../../../business/comment';
import {employee_actions} from '../../../business/employee';

const {formatTimestamp} = date_helpers;

const cx = classNames.bind(style_sheet);

const {orderStateTimelineEntries, commentIds, formatTimelineAttribution, isTerminalEntry} = order_helpers;

type OrderTimelineProps = { order: Order, comment_timeline_entries: Array<TimelineEntry>, fetchEmployees: () => void };

export class OrderTimeline extends PureComponent {
  props: OrderTimelineProps

  componentWillMount() {
    this.props.fetchEmployees();
  }

  render() {
    const {order, comment_timeline_entries} = this.props;
    const mixed_entries = [...orderStateTimelineEntries(order), ...comment_timeline_entries];
    const mixed_sorted_entries = _.sortBy(mixed_entries, [(entry) => moment.parseZone(entry.time)]);

    return (
      <div>
        {mixed_sorted_entries.map((entry, index) => {
          if (isTerminalEntry(entry) && index + 1 === mixed_sorted_entries.length) return <OrderTimelineEntry
            key={`${entry.time}${entry.type}`} entry={entry}/>;
          return [
            <OrderTimelineEntry key={`${entry.time}${entry.type}`} entry={entry}/>,
            <div key={`${entry.time}${entry.type}_separator`} className={style_sheet.separatorLine}/>
          ];
        })}
      </div>
    );
  }
}

type OrderTimelineEntryProps = { entry: TimelineEntry };
export const OrderTimelineEntry = ({entry}: OrderTimelineEntryProps) => {
  const dot_class = cx({
    entryDot: true,
    comment: entry.type === 'comment',
    delivered: entry.type === 'delivered_at',
    canceled: entry.type === 'canceled_at'
  });

  const determineIcon = (entry_type) => {
    switch (entry_type) {
      case 'comment':
        return 'note';
      case 'delivered_at':
        return 'check';
      case 'canceled_at':
        return 'cancel';
      default:
        return null;
    }
  };

  return (
    <div className={style_sheet.timelineEntry}>
      <div className={dot_class}>
        <MBIcon icon={determineIcon(entry.type)} color="mb_white"/>
      </div>
      <div className={style_sheet.timelineEntryText}>
        <p className={style_sheet.timelineEntryTitle}>{formatTimelineAttribution(entry)}</p>
        <p className={style_sheet.timelineEntryTimestamp}>{formatTimestamp(entry.time)}</p>
      </div>
    </div>
  );
};

const OrderTimelineSTP = () => {
  const findComments = Ent.find('comment');
  return (state, {order}) => {
    return {
      order: {...order, driver: state.employee.by_id[order.driver_id]},
      comment_timeline_entries: comment_helpers.commentTimelineEntries(findComments(state, commentIds(order)))
    };
  };
};
const OrderTimelineContainerDTP = {fetchEmployees: employee_actions.fetchEmployees};
const OrderTimelineContainer = connect(OrderTimelineSTP, OrderTimelineContainerDTP)(OrderTimeline);

export default OrderTimelineContainer;
