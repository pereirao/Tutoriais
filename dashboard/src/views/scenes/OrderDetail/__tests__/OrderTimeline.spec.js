import React from 'react';

import { comment_helpers } from '../../../../business/comment';
import comment_factory from '../../../../business/comment/__tests__/comment.factory';
import { order_helpers } from '../../../../business/order';
import order_factory from '../../../../business/order/__tests__/order.factory';
// NOTE: the named export isn't a connected component
import { OrderTimeline, OrderTimelineEntry } from '../OrderTimeline';

const { orderStateTimelineEntries } = order_helpers;

//TODO: make specs for all cases of Entry when they're laid out

it('renders OrderTimeline correctly', () => {
  const comment = comment_factory.build({id: 1});
  expect(shallow(
    <OrderTimeline
      comment_timeline_entries={comment_helpers.commentTimelineEntries([comment])}
      order={order_factory.build({comments: [comment]})} />
  )).toMatchSnapshot();
});

it('renders OrderTimelineEntry correctly', () => {
  const order = order_factory.build('paid');
  expect(shallow(<OrderTimelineEntry entry={orderStateTimelineEntries(order)[0]} />)).toMatchSnapshot();
});
