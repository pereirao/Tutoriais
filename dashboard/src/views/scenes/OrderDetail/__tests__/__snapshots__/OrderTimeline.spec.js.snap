// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders OrderTimeline correctly 1`] = `
<div>
  <OrderTimelineEntry
    entry={
      Object {
        "meta": Object {
          "name": "MF DOOM (employee)",
        },
        "time": "2017-01-01T00:00:00-05:00",
        "type": "comment",
      }
    }
  />
  <div
    className="separatorLine"
  />
</div>
`;

exports[`renders OrderTimelineEntry correctly 1`] = `
<div
  className="timelineEntry"
>
  <div
    className="entryDot"
  >
    <MBIcon
      color="mb_white"
      icon={null}
    />
  </div>
  <div
    className="timelineEntryText"
  >
    <p
      className="timelineEntryTitle"
    >
      Placed by Biggie Smalls (customer)
    </p>
    <p
      className="timelineEntryTimestamp"
    >
      12:00 AM on 01/02/2017
    </p>
  </div>
</div>
`;
